FirebaseSegmentation/Sources/SEGContentManager.m

```diff
@@ -0,0 +1,119 @@
+                 networkManager:[[SEGNetworkManager alloc] initWithFIROptions:options]];
+  });
+  return sharedInstance;
+}
+
+- (instancetype)initWithDatabaseManager:databaseManager networkManager:networkManager {
+  self = [super init];
+  if (self) {
+    // Initialize the database manager.
+    _databaseManager = databaseManager;
+
+    // Initialize the network manager.
+    _networkManager = networkManager;
+
+    NSAssert(_databaseManager != nil, @"Segmentation database could not be initialized");
+
+    // Load all data from the database.
+    [_databaseManager createOrOpenDatabaseWithCompletion:^(BOOL success, NSDictionary *result) {
+      self->_associationData = [result mutableCopy];
+    }];
+    // TODO(dmandar) subscribe to FIS notifications once integrated.
+  }
+  return self;
+}
+
+// TODO(dmandar) IID only supports default instance. Modify for FIS.
+- (FIRInstanceID *)instanceIDForApp:(NSString *)firebaseApp {
+  return [FIRInstanceID instanceID];
+}
+
+- (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallationID
+                                      firebaseApp:(NSString *)firebaseApp
+                                       completion:(SEGRequestCompletion)completionHandler {
+  // Get the latest instance identifier
+  if (![self instanceIDForApp:firebaseApp]) {
+    completionHandler(NO, @{@"ErrorDescription" : @"InstanceID SDK not available"});
+  }
+  __weak SEGContentManager *weakSelf = self;
+  [[FIRInstanceID instanceID] instanceIDWithHandler:^(FIRInstanceIDResult *_Nullable result,
+                                                      NSError *_Nullable error) {
+    SEGContentManager *strongSelf = weakSelf;
+    if (!strongSelf) {
+      completionHandler(NO, @{@"ErrorDescription" : @"Internal Error getting instance ID."});
+      return;
+    }
+
+    strongSelf->_instanceIdentifier = result.instanceID;
+    strongSelf->_instanceIdentifierToken = result.token;
+
+    NSMutableDictionary<NSString *, NSString *> *appAssociationData =
+        [[NSMutableDictionary alloc] init];
+    [appAssociationData setObject:customInstallationID forKey:kSEGCustomInstallationIdentifierKey];
+    [appAssociationData setObject:self->_instanceIdentifier
+                           forKey:kSEGFirebaseInstallationIdentifierKey];
+    [appAssociationData setObject:kSEGAssociationStatusPending forKey:kSEGAssociationStatusKey];
+    [strongSelf->_associationData setObject:appAssociationData forKey:firebaseApp];
+
+    // Update the database async.
+    [strongSelf->_databaseManager insertMainTableApplicationNamed:firebaseApp
+                                         customInstanceIdentifier:customInstallationID
+                                       firebaseInstanceIdentifier:strongSelf->_instanceIdentifier
+                                                associationStatus:kSEGAssociationStatusPending
+                                                completionHandler:nil];
```

Comment:
Can you please add a comment about why we don't care when the `completionHandler` is called here? It's unclear to me.

Reply:
updated comment.